### ISSUE TITLE
Inflexible API for language handlers

### ISSUE THREAD

**Issue by rchl** [sublimelsp/LSP#927]
Related to #897 and #926

`LanguageHandler` classes can implement `on_start` and `on_initialized` methods and it is enough for most servers but some need access to extra things which are not currently available:
 - `LSP-intelephense` needs access to Sublime window as it uses `window.extract_variables()` to get window variables. Currently it's getting window using `sublime.active_window()` but that is error-prone as there is no guarantee that active_window is the correct one.
 - `LSP-eslint` would need access to `WorkspaceFolders` in `on_initialized` as it needs to reply to `workspace/configuration` with workspace that matches the current file .
 - Tom is adding `can_start` method which is basically like `on_start` (both can prevent server from starting) but with different arguments 

I'd like us to have some flexible and extensible solution for these issues where we don't have to basically duplicate methods to add new arguments to make API backward-compatible.

**Comment by tomv564** [comment]
As we discussed in #897:
* Duplicating methods to add arguments is pretty ugly
* We can break the APIs once to create future-compatible APIs now.
* In practice people create ServerHandlers, not LanguageHandlers.

**If we create a new ServerHandler API, what do we take from LanguageHandler?**

* `on_start` sounds like an event handler but it actually decides if sessions can be started. I think `can_start` better reflects what the call is capable of.
  * Should we move all arguments into a CanStartArgs dict?
  * Is a list of startable folders a good response? Or move this into a dict as well? How can we easily represent yes/no/dontcare in this dict-form?
* `on_initialized` is still an event handler, is it a good name?
  * What do we want to add to an InitializedArgs dict?

**How did configuration management work out?**

This is in relation to the `config` member. I see packages like lsp-intelephense invented their own settings structure. It's unclear to me if those packages handled eg. project-specific overrides correctly and if it is all worth it?

Also there are requests to remove default configs for eg. eslint (even though endusers still may have references to that config in user settings).

Is there something about the way `config` is provided via LanguageHandler that's causing pain? Now is a good time to fix that as well.

**Comment by rchl** [comment]
> We can break the APIs once to create future-compatible APIs now.

If we could keep both APIs alive for a while, that would make the transition smoother.

> * I think `can_start` better reflects what the call is capable of.

That name sounds good to me.

> * Is a list of startable folders a good response?

Can we move that discussion to #926? I'd like to understand better the problem it's solving and how it relates to workspaces support. I'll raise a question there.

> `on_initialized` is still an event handler, is it a good name?
> * What do we want to add to an InitializedArgs dict?

I think it's a good name since it doesn't return anything.
For eslint I would need a list of workspace folders. @rwols mentioned somewhere else that it might work to add support for interpolating settings via `window.extract_variables()` but I'm not convinced (we could discuss in https://github.com/sublimelsp/LSP-eslint/pull/6).

> This is in relation to the `config` member. I see packages like lsp-intelephense invented their own settings structure.

Project-specific overrides are not handled currently in those packages AFAIK. It would be worth to support them.

> Also there are requests to remove default configs for eg. eslint (even though endusers still may have references to that config in user settings).

Not sure what are you referring to but the request to remove eslint settings was only due to the fact that we have `LSP-eslint` now. `LSP-eslint` still provides its own default settings. Is it related to this discussion though?

> Is there something about the way `config` is provided via LanguageHandler that's causing pain? Now is a good time to fix that as well.

LSP-intelephense wants to return some values from `config()` that are dependant on Sublime's `window` for a given file. At that point, there is no window yet though. At least not one that is specific to the project that the server will run on.

So it might be a good idea to split configuration into two parts. First that returns just `languages` configuration (`syntaxes`, `scopes` and `languageId`) for the LSP to know where it should initialize server and second that is retrieved when server is about to initialize and `window` is known already.
This would be also good as it would allow delaying some potentially expensive operations until server is about to initialize.
For example we could have `languages()` and `config(window)`

**Comment by tomv564** [comment]
Great response, thanks. Especially the comments about configs should lead to some improvements!

**Comment by rwols** [comment]
What we know so far:

1. ServerHandler knows what languages it applies to.
2. ServerHandler knows which completion selectors to apply (see https://github.com/sublimelsp/LSP/issues/933)

Both of these bullet points are independent of the instance, so they can be class methods.
Based on this, the interface should include

```
class ServerHandler:

    @classmethod
    def languages(cls) -> Optional[List[LanguageConfig]]: ...

    @classmethod
    def completion_selectors(cls) -> Optional[dict]: ...
```

**Comment by rchl** [comment]
Should `languages` be required? Would there be a use-case for a handler without those?

**Comment by predragnikolic** [comment]
@rchl `languages` config key was added later during the LSP development.
as it allowed the `clients` settings key to write client configuration in a nicer way.

The perfect example for this is the CSS language server:
Before `languages` were added you would have to create a config like this:
```
{
  "clients": 
  {
    "vscode-css":
    {
      "command": ["node", "/usr/local/lib/node_modules/vscode-css-languageserver-bin/cssServerMain.js", "--stdio"],
      "scopes": ["source.css"],
      "syntaxes": ["Packages/CSS/CSS.sublime-syntax"],
      "languageId": "css"
    },
    "vscode-scss":
    {
      "command": ["node", "/usr/local/lib/node_modules/vscode-css-languageserver-bin/cssServerMain.js", "--stdio"],
      "scopes": ["source.scss"],
      "syntaxes": ["Packages/Sass/Syntaxes/SCSS.sublime-syntax"],
      "languageId": "scss"
    },
    "vscode-less":
    {
      "command": ["node", "/usr/local/lib/node_modules/vscode-css-languageserver-bin/cssServerMain.js", "--stdio"],
      "scopes": ["source.css.less"],
      "syntaxes": ["Packages/LESS/LESS.sublime-syntax"],
      "languageId": "less"
    },
    "vscode-sass":
    {
      "command": ["node", "/usr/local/lib/node_modules/vscode-css-languageserver-bin/cssServerMain.js", "--stdio"],
      "scopes": ["source.sass"],
      "syntaxes": ["Packages/Sass/Syntaxes/Sass.sublime-syntax"],
      "languageId": "sass"
    },
  }
}
```

Later when `languages` was added, you could rewrite your config like this:
 ```
{
  "clients": {
    "vscode-css": {
      "command": ["node", "/usr/local/lib/node_modules/vscode-css-languageserver-bin/cssServerMain.js", "--stdio"],
      "languages": {
        "css": {
          "scopes": ["source.css"],
          "syntaxes": ["Packages/CSS/CSS.sublime-syntax"],
          "languageId": "css"
        },
        "scss": {
          "scopes": ["source.sass"],
          "syntaxes": ["Packages/Sass/Syntaxes/Sass.sublime-syntax"],
          "languageId": "scss"
        },
        "sass": {
          "scopes": ["source.sass"],
          "syntaxes": ["Packages/Sass/Syntaxes/Sass.sublime-syntax"],
          "languageId": "sass"
        },
        "less": {
          "scopes": ["source.css.less"],
          "syntaxes": ["Packages/LESS/LESS.sublime-syntax"],
          "languageId": "less"
        }
      }
    },
  }
}
```

**Comment by rchl** [comment]
Yes, but I was referring to type annotation being optional.

**Comment by rwols** [comment]
You're right, it makes no sense to have `None` returned from such a classmethod, as then the LanguageHandler would apply to no view and thus would never be started.

**Comment by ayoub-benali** [comment]
Another feature request: 
In the case of metals I need to fill the "experimental" field (part of `ClientCapabilities`) sent during the `initialize` request with these [values](https://scalameta.org/metals/docs/editors/new-editor.html#experimental) 

So `on_initialized` is too late in the process. 
Maybe allow LanguageHandler to define capabilities that LSP would merge with the default one ?

**Comment by rwols** [comment]
Based on what I've been playing around with in `exit-on-exit` branch:

- Cyclic references caused by keeping `Client` around in a bunch of event handlers causes the subprocess to not be terminated on application exit. Moreover the callbacks in the `Transport` are a strong reference to `Client`, while at the same time `Client` has a strong reference to its `Transport` -> cyclic reference.
- Cyclic references caused by putting lambda's and bound methods in the `self._request_handlers` and `self._notification_handlers` dicts in `Client` causes the subprocess to not be terminated on application exit.
- I see no real need to separate the idea of a `Client` and a `Session`. I've made `Session` inherit from `Client` from now, but it is becoming more clear that that is also pointless.

In order to remove cyclic references I've stolen the idea of python-language-server to delegate notification/request handlers to actual methods of the Session.

For instance, to handle the `workspace/configuration` request you would write a method in `Session` called

```python
def m_workspace_configuration(self, params: Any, request_id: Any) -> None:
    ...
```

I succeeded in removing the cyclic references and tested the branch on Windows. The subprocesses are correctly terminated now when closing SublimeText. No need for the new `on_exit` event handler from `sublime_plugin.EventListener`!

While this idea is cool, this breaks `LanguageHandler` entirely as it won't be able to handle custom requests/notifications anymore. But because removing cyclic references is so important IMO, we should introduce some breakage for ST4. As I said before it makes sense to me to let a "language" or "server" handler to inherit from Session. It's a custom Session. Below is a design I've been playing around with:

```python
class VscodeJsonLanguageServer(Session):

    @classmethod
    def name(cls) -> str:
        return 'vscode-json-languageserver'

    @classmethod
    def needs_update_or_installation(cls) -> bool:
        """
        If this custom session manages its own server binary, this is the place to check whether that server binary
        requires an update or an installation.
        """
        return False

    @classmethod
    def install_or_update(cls) -> None:
        """
        If this custom session manages its own server binary, install or update the binary here. Do not spawn threads.
        This function will run outside the UI thread already.
        """
        pass

    @classmethod
    def provide_configuration(cls) -> Optional[ClientConfig]:
        """
        Provide a configuration so that the user doesn't have to write anything in a .sublime-settings file.
        The configuration's name MUST be the same as the name() classmethod above.
        """
        return None

    @classmethod
    def adjust_configuration(cls, configuration: ClientConfig) -> ClientConfig:
        """
        Given a configuration written by the user in a .sublime-settings file or a .sublime-project, adjust the
        configuration here for an easier/better experience.
        """
        return configuration

    @classmethod
    def can_start(cls, window: sublime.Window, initiating_view: sublime.View,
                  workspace_folders: List[WorkspaceFolder], configuration: ClientConfig) -> Optional[str]:
        """
        LSP is about to create an instance of this session. Is it possible to start the session? If not, return a string
        describing the reason why it's not possible to start the session.
        """
        return None

    def on_initialized(self) -> None:
        """
        Called after the language server returned a response from the initialize request. The capabilities are filled
        in and the workspace folders have been adjusted.
        """
        pass

    def on_shutdown(self) -> None:
        """
        Called just before sending the shutdown request.
        """
        pass

    def m_vscode_content(self, params: Any, request_id: Any) -> None:
        """
        Handles the vscode/content request.
        """
        name = params.split('/')[-1]
        schema_resource = ResourcePath('Packages', PACKAGE_NAME, 'schemas', '{}.json'.format(name))
        self.send_response(Response(request_id, schema_resource.read_text()))

```
I hope you agree that the request handler for `vscode/content` is beautifully straightforward.

The `name()` classmethod can be used as a key in a global dict of `str -> Type[Session]`. When LSP is about to start a session, the `name` of the `ClientConfig` is then used to look up the correct derived `Session`. So in this case it would look like

```python

derived_sessions = {}  # type: Dict[str, Type[Session]]

class WindowManager:
    def _start_client(self, config: ClientConfig):
        session_class = derived_sessions.get(config.name, Session)
        config = session_class.adjust_configuration(config)
        reason = session_class.can_start(...)
        if isinstance(reason, str):
            # cannot start!
        ...
```

**Comment by ayoub-benali** [comment]
> No need for the new on_exit event handler from sublime_plugin.EventListener

This is really cool and in general I really like the new API, it is much cleaner.

> we should introduce some breakage for ST4.

Why specifically ST4? The API could be changed after a specific LSP version and since most language handlers are in https://github.com/sublimelsp  we could prepare them in advance, no ?

**Comment by predragnikolic** [comment]
> In the case of metals I need to fill the "experimental" field (part of ClientCapabilities) sent during the initialize request with these values
So on_initialized is too late in the process.

Can't we already do this?
@ayoub-benali can you pass this
```
init_options={
  "experimental": {
    "treeViewProvider": boolean,
    "debuggingProvider": boolean,
    "decorationProvider": boolean,
    "inputBoxProvider": boolean,
    "quickPickProvider": boolean,
    "didFocusProvider": boolean,
    "slowTaskProvider": boolean,
    "executeClientCommandProvider": boolean,
    "doctorProvider": "json" | "html",
    "statusBarProvider": "on" | "off" | "show-message" | "log-message"
  }
}
```
here, instead of an empty dict?
https://github.com/scalameta/metals-sublime/blob/40f63a57e66947d572c30f58d702f292a409382a/plugin.py#L70

**Comment by ayoub-benali** [comment]
experimental capabilities isn't part of initialzation options. But that limitation was handled here https://github.com/sublimelsp/LSP/pull/971

**Comment by predragnikolic** [comment]
I didn't know that. 
I didn't event know that "experimental" is a part of the spec. Thanks for clarifying

**Comment by rchl** [comment]
So as I understand both `needs_update_or_installation` and `install_or_update` would be run by LSP and not the package? If so then `needs_update_or_installation` should also mention that.

Also, `needs_update_or_installation` should probably also run in a separate thread as the process of if update is needed can be expensive and trigger file access.

And `can_start` should be aware of those above so that the package doesn't have to keep state required to return True or False here. I mean the state related to installation/updating.

**Comment by rwols** [comment]
> So as I understand both needs_update_or_installation and install_or_update would be run by LSP and not the package?

That would be my idea, yes.

> If so then needs_update_or_installation should also mention that.

Good point.

> Also, needs_update_or_installation should probably also run in a separate thread as the process of checking if update is needed can be expensive and trigger file access.

Hmm. This would be somewhat annoying. We would have to start a py thread just for checking whether an update is needed. That means starting an extra short-lived py thread every time the user opens ST. This is IMO bad design.

We could try running it on the ST async thread, but I worry this will deadlock if we're already running on the async thread.

How expensive is it really to check whether an update is needed? Can you give some concrete examples?

Also, once we move to Python 3.8, we can change `install_or_update` to an `async` function and we won't need `needs_update_or_installation` anymore. I view `needs_update_or_installation` as a temporary measure for Python 3.3.

> And can_start should be aware of those above so that the package doesn't have to keep state required to return True or False here. I mean the state related to installation/updating.

This can be a bunch of class-level variables? There's only ever going to be at most one server binary.
The way I see it is that `needs_update_or_installation` is always called before `can_start`, so one can put the necessary class-level info in `needs_update_or_installation`.

**Comment by rchl** [comment]
> How expensive is it really to check whether an update is needed? Can you give some concrete examples?

For the current nodejs servers we are doing:
- get cache path with `sublime.cache_path()`
- read contents of the `package.json` files from the package and from the cache directory
- compare as strings

So it's basically IO to read content of both files. And we are doing it synchronously right now so I guess it's fine. Sorry for panicking. ;)

> The way I see it is that `needs_update_or_installation` is always called before `can_start`, so one can put the necessary class-level info in `needs_update_or_installation`.

OK, that should work. :)

There is one other thing, currently, since installation of server is async, `can_start` will return `False` for views while server is installing. Those views have to be re-opened to make installed server start for those views. New API should have that in mind and automatically re-attempt to init once server is installed (since it will know when that happens).

**Comment by rwols** [comment]
> New API should have that in mind and automatically re-attempt to init once server is installed (since it will know when that happens).

Yep that's the idea. If the base package and the helper package work together we can make it work so that server installation happens just-before starting the session.

`needs_update_or_installation` is only required to exist to check whether to create a short-lived py thread. It should in most cases return `False` so that we're not spawning this thread most of the time. Again once we move to Python 3.8 this function can disappear because running `async` functions is always cheap (if the event loop is always running).

**Comment by rwols** [comment]
Fixed by #979
